File: C/scanner.c


#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <math.h>
#include <string.h>
#include "defines.h"

typedef struct Token{
  int type;
  char* lexeme;
  int line;

  char* string_literal;
  int string_length;
  double num_literal;
  struct Token* next;//to make a linked list of tokens
  struct Token* prev;
  //oh god what have I done
}Token;

typedef struct Scanner{
  int start;
  int current;
  int line;

  char* source;
  struct Token* tokens;
  struct Token* tail;
  int num_chars_in_input;
}Scanner;

void print_all_tokens(struct Scanner* S){
  printf("--------------------------------------------------------------------\n");
  struct Token* t = S->tokens;
  while(t != NULL){
    printf("Token of type: %d     ",t->type);
    if(t->type == NUMBER){
      printf("%f",t->num_literal);
    }
    if(t->type == STRING){
      for(int i=0; i<t->string_length; i++){
        printf("%c",(t->string_literal)[i]);
      }
    }
    if(t->type == VARIABLE){
      printf("(var) ");
    }
    if((t->type <=116) & (t->type >=100)){
      for(int i=0; i<t->string_length; i++){
        printf("%c",(t->string_literal)[i]);
      }
    }

    printf("\n");
    t = t->next;
  }
    printf("--------------------------------------------------------------------\n");
}

bool is_alphabetic(char c){
  if((c>=97 & c<=122) | (c >=65 & c<=90) | (c==95)){
    return true;
  }
  return false;
}

bool is_digit(char c){
  if(c>=48 & c<=57){
    return true;
  }
  return false;
}

bool is_alphanumeric(char c){
  return (is_digit(c) | is_alphabetic(c));
}

struct Token* new_token(int type, char* lexeme, int line, char* string_literal, int string_length, double num_literal){
  struct Token* T = malloc(1*sizeof(Token));
  T->type = type;
  T->lexeme = lexeme;
  T->line = line;
  T->string_literal = string_literal;
  T->string_length = string_length;
  T->num_literal = num_literal;
  T->next = NULL;
  T->prev = NULL;
  return T;
}

struct Scanner* new_scanner(char* source){
  struct Scanner* S = calloc(1,sizeof(Scanner));
  S->source = source;
  S->start = 0;
  S->current = 0;
  S->line = 1;
  return S;
}

void error_message(int line, char* message){
  printf("ERROR : line %d : %s\n",line,message);
}


bool is_at_end(struct Scanner* S){
  return(S->current >= S->num_chars_in_input);
}

void free_tokens(struct Scanner* S){
  struct Token* T = S->tokens;
  while(T != NULL){
    struct Token* tmp = T->next;
    free(T->lexeme);
    free(T->string_literal);
    free(T);
    T = tmp;
  }
}

void tokens_append(struct Scanner* S,struct Token* t){
  if(S->tail != NULL){
    S->tail->next = t;
    t->prev = S->tail;
    S->tail = t;
  }
  else{
    S->tokens = t;
    S->tail = t;
  }

}

void add_token(struct Scanner* S, int type, char* string_literal, int string_length, double num_literal){
  int len = (S->current) - (S->start);
  // printf("len: %d\n",len);
  if(len < 0){
    printf("invalid length\n");
    return;
  }
  char* text = malloc(len*sizeof(char));
  int idx = 0;
  for(int i=S->start; i<S->current; i++){
      text[idx] = (S->source)[i];
      idx++;
  }
  struct Token* nt = new_token(type,text,S->line,string_literal,string_length,num_literal);
  tokens_append(S,nt);
}



char advance(struct Scanner* S){
  (S->current)++;
  return (S->source)[(S->current)-1];
}

bool match(struct Scanner* S, char expected){
    if(is_at_end(S)){
      return false;
    }
    if((S->source)[S->current] != expected){
      return false;
    }
    (S->current)++;
    return true;
}

char peek(struct Scanner* S){
  if(is_at_end(S)){
    return '\0';
  }
  return (S->source)[S->current];
}

char peek_next(struct Scanner* S){
  if((S->current)+1 >= S->num_chars_in_input){
    return 0;
  }
  return (S->source)[S->current+1];
}

void scan_number(struct Scanner* S){

  while(is_digit(peek(S))){
    advance(S);
  }

  if((peek(S) == 46) & (is_digit(peek_next(S)))){
    advance(S);
    while(is_digit(peek(S))){
      advance(S);
    }
  }
  //from start to current contains the chars for the value
  int num_digits = (S->current) - (S->start);
  // printf("%d digits\n",num_digits);
  int i=S->start;
  for(i=S->start; i<S->current; i++){
    if((S->source)[i] == 46){
      //hit a decimal point
      break;
    }
  }

  //upper from start to i
  //lower from i+1 to current
  double output = 0;
  int power = 0;
  int d;
  for(d=i-1; d>(S->start)-1; d--){
    int digit = ((S->source)[d])-48;
    // printf("encountered %d\n",digit);
    output += (digit * pow(10,power));
    power++;
  }
  // printf("upper\n");

// printf("d: %d     start %d       current %d     i %d\n",d,S->start,S->current,i);



  power = -1;
  for(int d=i+1; d<(S->current); d++){
    int digit = ((S->source)[d])-48;
    output += (digit * pow(10,power));
    power--;
  }

  // printf("lower\n");
  // printf("%lf\n",output);

  add_token(S,NUMBER,NULL,0,output);

}

void scan_string(struct Scanner* S){
  while(peek(S) != 34 & (is_at_end(S) == false)){
    if(peek(S) == 10){
      (S->line)++;
    }
    advance(S);
  }

  if(is_at_end(S)){
    error_message(S->line,"unterminated string");
    return;
  }

  advance(S);

  int len = ((S->current)-1) - ((S->start)+1);
  char* text = malloc(len*sizeof(char));
  int idx = 0;
  for(int i=(S->start)+1; i<(S->current)-1; i++){
    text[idx] = (S->source)[i];
    idx++;
  }

  // for(int i=0; i<len; i++){
  //   printf("%c",text[i]);
  // }
  // printf("\n");

  add_token(S,STRING,text,len,0);

}

void scan_identifier(struct Scanner* S){
  while(is_alphanumeric(peek(S))){
    advance(S);
  }
  int len = (S->current) - (S->start);
  char* text = malloc((len+1)*sizeof(char));
  int idx = 0;
  for(int i=S->start; i<S->current; i++){
      text[idx] = (S->source)[i];
      idx++;
  }
  text[len] = '\0';//to make it a string!

  printf("text 0:  %d\n",text[0]);

  int type;
  if((text != NULL) & (text[0] != '\0')){
    if(strcmp(text,"and") == 0){
      type = KEYWORD_AND;
    }
    else if(strcmp(text,"class") == 0){
      type = KEYWORD_CLASS;
    }
    else if(strcmp(text,"else") == 0){
      type = KEYWORD_ELSE;
    }
    else if(strcmp(text,"false") == 0){
      type = KEYWORD_FALSE;
    }
    else if(strcmp(text,"for") == 0){
      type = KEYWORD_FOR;
    }
    else if(strcmp(text,"fun") == 0){
      type = KEYWORD_FUN;
    }
    else if(strcmp(text,"if") == 0){
      type = KEYWORD_IF;
    }
    else if(strcmp(text,"nil") == 0){
      type = KEYWORD_NIL;
    }
    else if(strcmp(text,"or") == 0){
      type = KEYWORD_OR;
    }
    else if(strcmp(text,"print") == 0){
      type = KEYWORD_PRINT;
    }
    else if(strcmp(text,"return") == 0){
      type = KEYWORD_RETURN;
    }
    else if(strcmp(text,"super") == 0){
      type = KEYWORD_SUPER;
    }
    else if(strcmp(text,"this") == 0){
      type = KEYWORD_THIS;
    }
    else if(strcmp(text,"true") == 0){
      type = KEYWORD_TRUE;
    }
    else if(strcmp(text,"var") == 0){
      type = KEYWORD_VAR;
    }
    else if(strcmp(text,"while") == 0){
      type = KEYWORD_WHILE;
    }
    else{
      type = VARIABLE;
    }


    add_token(S,type,text,len,0);

  }

}

void scan_token(struct Scanner* S){
  char c = advance(S);
  switch(c){
    case 40:
      add_token(S,LPAREN,NULL,0,0);
      break;
    case 41:
      add_token(S,RPAREN,NULL,0,0);
      break;
    case 125:
      add_token(S,RCURLY,NULL,0,0);
      break;
    case 123:
      add_token(S,LCURLY,NULL,0,0);
      break;
    case 91:
      add_token(S,LBRACKET,NULL,0,0);
      break;
    case 93:
      add_token(S,RBRACKET,NULL,0,0);
    case 44:
      add_token(S,COMMA,NULL,0,0);
      break;
    case 46:
      add_token(S,DOT,NULL,0,0);
      break;
    case 45:
      add_token(S,MINUS,NULL,0,0);
      break;
    case 43:
      add_token(S,PLUS,NULL,0,0);
      break;
    case 59:
      add_token(S,SEMICOLON,NULL,0,0);
      break;
    case 42:
      add_token(S,STAR,NULL,0,0);
      break;
    case 33:
      if(match(S,61)){
          add_token(S,NOTEQUAL,NULL,0,0);
      }
      else{
        add_token(S,EXLAMATION,NULL,0,0);
      }
      break;
    case 61:
      if(match(S,61)){
        add_token(S,EQUALEQUAL,NULL,0,0);
      }
      else{
        add_token(S,EQUAL,NULL,0,0);
      }
      break;
    case 60:
      if(match(S,61)){
        add_token(S,LESSEQUAL,NULL,0,0);
      }
      else{
        add_token(S,LESSTHAN,NULL,0,0);
      }
      break;
    case 62:
      if(match(S,61)){
          add_token(S,GREATEREQUAL,NULL,0,0);
      }
      else{
        add_token(S,GREATERTHAN,NULL,0,0);
      }
      break;
    case 32:
      break;
    case 9:
      break;
    case 13:
      break;
    case 10:
      (S->line)++;
      break;
    default:
      if(c==34){
        scan_string(S);
      }
      else if(is_digit(c)){
        scan_number(S);
      }
      else if(is_alphabetic(c)){
        scan_identifier(S);
      }
      else{
        error_message(S->line,"unexpected character");
      }
      break;


  }


}

struct Token* scan_tokens(struct Scanner* S){
  printf("scanning tokens\n");
  while(!is_at_end(S)){
    S->start = S->current;
    scan_token(S);
  }
  tokens_append(S,new_token(EOFTOKEN,0,0,NULL,0,0));
}

File: scanner.py

class Scanner:
    def __init__(self,byte_arr):
        self.start = 0
        self.current = 0
        self.line = 1

        self.byte_arr = byte_arr#called 'source' in book
        self.tokens = []

        self.keywords = [b'and',b'class',b'else',b'false',b'for',b'fun',b'if',b'nil',b'or',b'print',b'return',b'super',b'this',b'true',b'var',b'while']


    def print_byte_arr(self):
        for i in range(0,len(self.byte_arr)):
            print(self.byte_arr[i])

    def is_at_end(self):
        return (self.current >= len(self.byte_arr))

    def scan_tokens(self):
        print("scanning tokens")
        while(self.is_at_end() == False):
            self.start = self.current
            self.scan_token()


        (self.tokens).append(Token("EOF","",None,None))#I think?
        #in book it is (EOF,"","NULL","line")
        return(self.tokens)
        #must return some list of token objects

    def scan_token(self):
        c = self.advance()
        if(c==40):
            self.add_token('(',None)
        elif(c==41):
            self.add_token(')',None)
        elif(c==123):
            self.add_token('{',None)
        elif(c==125):
            self.add_token('}',None)
        elif(c==44):
            self.add_token(',',None)
        elif(c==46):
            self.add_token('.',None)
        elif(c==45):
            self.add_token('-',None)
        elif(c==43):
            self.add_token('+',None)
        elif(c==59):
            self.add_token(';',None)
        elif(c==42):
            self.add_token('*',None)
        elif(c==33):
            if(self.match(61)):
                self.add_token('!=',None)
            else:
                self.add_token('!',None)
        elif(c==61):
            if(self.match(61)):
                self.add_token('==',None)
            else:
                self.add_token('=',None)
        elif(c==60):
            if(self.match(61)):
                self.add_token('<=',None)
            else:
                self.add_token('<',None)
        elif(c==62):
            if(self.match(61)):
                self.add_token('>=',None)
            else:
                self.add_token('>',None)
        elif(c==47):#comment
            if(self.match(47)):
                while(self.peek() != '\n' and (self.is_at_end() == False)):
                    self.advance()
            else:
                self.add_token('/',None)
        elif(c==32):#space
            pass
        elif(c==9):#tab
            pass
        elif(c==13):#carriage return
            pass
        elif(c==10):#newline
            self.line += 1
        elif(c==34):#double quotes
            self.scan_string()
        elif(self.is_digit(c)):
            self.scan_number()
        elif(self.is_alphabetic(c)):
            self.scan_indentifier()
        else:
            Z.error_message(self.line,"unexpected character")

    def scan_indentifier(self):
        while(self.is_alphanumeric(self.peek())):
            self.advance()

        text = self.byte_arr[self.start:self.current:1]
        text = bytes(text)
        print(text)
        if(text in self.keywords):
            print("keyword detected")
            type = text
        else:
            type = 'identifier'

        self.add_token(type,None)

    def is_alphabetic(self,c):
        if((c >= 97 and c <= 122) or (c >= 65 and c<=90) or (c == 95)):
            return True
        else:
            return False

    def is_alphanumeric(self,c):
        return (self.is_alphabetic(c) or self.is_digit(c))

    def scan_number(self):
        while(self.is_digit(self.peek())):
            self.advance()

        if(self.peek() == 46 and self.is_digit(self.peek_next())):
            self.advance()

            while(self.is_digit(self.peek())):
                self.advance()

        value = self.byte_arr[self.start:self.current:1]
        value = self.array_to_double(value)
        self.add_token('number',value)


    def array_to_double(self,v):
        num_digits = len(v)
        # print(v)
        for i in range(0,num_digits):
            if(v[i]==46):
                break
        upper = v[0:i:1]
        lower = v[i+1:num_digits:1]
        # print(upper)
        # print(lower)

        output = 0
        power = 0
        for d in reversed(upper):
            d = d-48
            output += (d*(10**power))
            power += 1

        power = -1
        for d in lower:
            d = d-48
            output += (d*(10**power))
            power -= 1

        # print(output)
        return output


    def is_digit(self,c):
        return ((c >= 48) and (c <= 57))

    def advance(self):
        self.current = self.current+1
        return (self.byte_arr)[self.current-1]

    def match(self,expected):
        if(self.is_at_end()):
            return False
        if((self.byte_arr)[self.current] != expected):
            return False

        self.current += 1
        return True

    def peek(self):
        if(self.is_at_end()):
            return '\0'
        return (self.byte_arr)[self.current]

    def peek_next(self):
        if(self.current+1 >= len(self.byte_arr)):
            return 0
        return self.byte_arr[self.current+1]

    def scan_string(self):
        while(self.peek() != 34 and (self.is_at_end() == False)):
            if(self.peek() == 10):
                self.line += 1
            self.advance()

        if(self.is_at_end()):
            Z.error_message(self.line,"unterminated string")
            return

        self.advance()

        value = self.byte_arr[self.start+1:self.current-1:1]
        value_bytes = bytes(value)
        print(value_bytes)
        # for i in range(0,len(value)):
        #     integer_val = value[i]
        #     byte_val = byte(integer_val)
        #     value_bytes[i] = byte_val

        # b''.join(value_bytes).decode('utf-8')

        self.add_token('string',value_bytes)
    # def add_token(type):
    #     add_token(type,None)

    def add_token(self,type,literal):
        # print("adding token")
        text = (self.byte_arr)[self.start:self.current:1]
        (self.tokens).append(Token(type,text,literal,self.line))




class Token:
    def __init__(self,type,lexeme,literal,line):
        self.type = type
        self.lexeme = lexeme
        self.literal = literal
        self.line = line
